go/vt/sidecardb/sidecardb.go

```diff
@@ -0,0 +1,373 @@
+
+	if err = si.loadExistingTables(); err != nil {
+		return err
+	}
+
+	for _, table := range vtTables {
+		if err := si.createOrUpgradeTable(table); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (si *vtSchemaInit) createVTDatabase() error {
+	_, err := si.exec(si.ctx, CreateVTDatabaseQuery, 1, false, false)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	log.Infof("createVTDatabase: %s", CreateVTDatabaseQuery)
+	return nil
+}
+func (si *vtSchemaInit) doesVTDatabaseExist() (bool, error) {
+	rs, err := si.exec(si.ctx, ShowVTDatabasesQuery, 2, false, false)
+	if err != nil {
+		log.Error(err)
+		return false, err
+	}
+
+	switch len(rs.Rows) {
+	case 0:
+		log.Infof("doesVTDatabaseExist: not found")
+		return false, nil
+	case 1:
+		log.Infof("doesVTDatabaseExist: found")
+		return true, nil
+	default:
+		log.Errorf("found too many rows for _vt: %d", len(rs.Rows))
+		return false, fmt.Errorf("found too many rows for _vt: %d", len(rs.Rows))
+	}
+}
+
+// sets db of current connection, returning the currently selected database
+func (si *vtSchemaInit) setCurrentDatabase(dbName string) (string, error) {
+	rs, err := si.exec(si.ctx, SelectCurrentDatabaseQuery, 1, false, false)
+	if err != nil {
+		return "", err
+	}
+	if rs == nil || rs.Rows == nil { // we get this in tests
+		return "", nil
+	}
+	currentDB := rs.Rows[0][0].ToString()
+	if currentDB != "" { // while running tests we can get currentDB as empty
+		_, err = si.exec(si.ctx, fmt.Sprintf("use %s", dbName), 1000, false, false)
+		if err != nil {
+			return "", err
+		}
+	}
+	return currentDB, nil
+}
+
+// gets existing schema of table
+func (si *vtSchemaInit) getCurrentSchema(tableName string) (string, error) {
+	var currentTableSchema string
+
+	rs, err := si.exec(si.ctx, fmt.Sprintf(ShowCreateTableQuery, tableName), 1, false, false)
+	if err != nil {
+		log.Errorf("Error getting _vt table schema for %s: %+v", tableName, err)
+		return "", err
+	}
+	if len(rs.Rows) > 0 {
+		currentTableSchema = rs.Rows[0][1].ToString()
+	}
+	return currentTableSchema, nil
+}
+
+// remove the extra DEFAULT CHARSET which we get from "show create table"
+func stripCharset(schema string) string {
+	ind := strings.Index(schema, "DEFAULT CHARSET")
+	if ind <= 0 {
+		return schema
+	}
+	return schema[:ind]
+}
+
+// finds diff that needs to be applied to current table schema to get the desired one. Will be an empty string if they match.
+// could be a create statement if the table does not exist or an alter if table exists but has a different schema
+func (si *vtSchemaInit) findTableSchemaDiff(current, desired string) (string, error) {
+	// todo: can we fix schemadiff to avoid the need for this?
+	current = stripCharset(current)
+
+	hints := &schemadiff.DiffHints{}
+	diff, err := schemadiff.DiffCreateTablesQueries(current, desired, hints)
+	if err != nil {
+		return "", err
+	}
+
+	tableAlterSQL := diff.CanonicalStatementString()
+	if strings.Contains(tableAlterSQL, "ALTER") {
```

Comment:
if anything, please use `if strings.HasPrefix(tableAlterSQL, "ALTER")`.

But better yet: the returned `diff` value is a `AlterTableEntityDiff`. It is either `nil` (no diff) or it is an `ALTER TABLE`. So looking for the word "ALTER" is superfluous. Just check if it isn't `nil`.